[RFC] x86: 64: add additional SAS / HBA card support

wigyori opened the pull request:
Certain users would like to use OpenWrt on their "brand" servers, where the rootdisk/array usually resides on disks behind a RAID card.

Support is added for:

 - hpsa (HP Smartarray)
 - ips (IBM ServeRAID)
 - megaraid (SAS / some legacy cards)
 - mpi3mr (SAS24G cards)
 - mpt3sas (SAS cards)

Kernel sizes compared:

Current:
6099968 Jan 30 18:13 openwrt-x86-64-generic-kernel.bin

New:
6358016 Jan 31 15:19 openwrt-x86-64-generic-kernel.bin

The 260kByte increase on an x86_64 box - which usually have 512Mb RAM or much more - is close to negligible, while it adds support for a significant portion of these servers / HBAs. This change should be limited to the 64-bit subtarget only. No commandline tool will be added this time, as: a) the cards can be configured via their BIOS at boot-time, b) licensing for those needs to be thoroughly reviewed.

Link: https://github.com/openwrt/openwrt/issues/17772

nasbdh9 commented:
Considering that running openwrt on a hardware raid controller/standalone hba is a rare case, my suggestion is to add a kernel option for these, and users in need can turn on these options to compile the kernel that needs these support.

Please refer to: https://github.com/openwrt/openwrt/blob/main/config/Config-kernel.in

wigyori commented:
> Considering that running openwrt on a hardware raid controller/standalone hba is a rare case, my suggestion is to add a kernel option for these, and users in need can turn on these options to compile the kernel that needs these support.
> 
> And these controllers are not only supported on x86 platforms, but also on arm/ppc/mips.

I wouldn't call that a rare case - the user in the referenced issue wasn't the first to ask about this. It's a more rare case to see an HPSA in a powerpc board though, which is why I was thinking about limiting this to x86_64 - just like the recent KERNEL_DCB support discussions.

Adding a kernel option for this would work, but I would more like to see this built into the snapshot/release images.

nightcoffee commented:
> Considering that running openwrt on a hardware raid controller/standalone hba is a rare case, my suggestion is to add a kernel option for these, and users in need can turn on these options to compile the kernel that needs these support.
> 
> And these controllers are not only supported on x86 platforms, but also on arm/ppc/mips.
> 
> Please refer to: https://github.com/openwrt/openwrt/blob/main/config/Config-kernel.in

Since a large number of old-model servers being phased out from data centers have become a cost-effective and reliable option, I believe running OpenWrt on these SAS or HBA cards would be a common choice (as these servers usually cannot directly utilize built-in SATA controllers). However, manually compiling the kernel would result in incompatibility with the official repositories.

hauke commented:
I would prefer to compile this as modules if possible, then we can use this also on other targets. The ARMSR target might get connected to such hardware too.

wigyori commented:
> I would prefer to compile this as modules if possible, then we can use this also on other targets. The ARMSR target might get connected to such hardware too.

I don't have an issue with packaging these (already did for the mpt3sas in staging), but to use them as rootdisks you need these in the kernel.

wigyori committed:
x86: 64: add additional SAS / HBA card support

Certain users would like to use OpenWrt on their brand servers, where
the rootdisk/array usually resides on disks behind a RAID card.

Support is added for:

 - hpsa (HP Smartarray)
 - ips (IBM ServeRAID)
 - megaraid (SAS / some legacy cards)
 - mpi3mr (SAS24G cards)
 - mpt3sas (SAS cards)

Kernel sizes compared:

Current:
6099968 Jan 30 18:13 openwrt-x86-64-generic-kernel.bin

New:
6358016 Jan 31 15:19 openwrt-x86-64-generic-kernel.bin

The 260kByte increase on an x86_64 box - which usually have 512Mb RAM or much
more - is close to negligible, while it adds support for a significant portion
of these servers / HBAs. This change should be limited to the 64-bit
subtarget only. No commandline tool will be added this time, as:
a) these can be configured via their BIOS at boot-time,
b) licensing for those would need to be thoroughly reviewed.

Link: https://github.com/openwrt/openwrt/issues/17772

Signed-off-by: Zoltan HERPAI <wigyori@uid0.hu>
